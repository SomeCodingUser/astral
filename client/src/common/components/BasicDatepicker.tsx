import { FC, useEffect, useState } from 'react'
import Datepicker from 'tailwind-datepicker-react'
import dayjs from 'dayjs'
import utc from 'dayjs/plugin/utc'
import { CalendarIcon } from '@heroicons/react/24/outline'
import { IOptions } from 'tailwind-datepicker-react/types/Options'

dayjs.extend(utc)

type Props = {
  value?: string | undefined
  onChange?: (selectedDate: Date) => void
  minDate?: Date
}

const defaultOptions = {
  theme: {
    background: 'bg-white dark:bg-[#1E254E] dark:text-white',
    disabledText: 'text-gray-500/75 dark:text-gray-400/75',
    input: 'py-3 w-full rounded-[20px] bg-white dark:bg-[#1E254E] dark:text-white',
    todayBtn: '',
    clearBtn: '',
    icons: '',
    text: '',
    inputIcon: '',
    selected: '',
  },
}

const BasicDatepicker: FC<Props> = ({ onChange, value, minDate }) => {
  const [show, setShow] = useState<boolean>(false)
  const [options, setOptions] = useState<IOptions>(defaultOptions)

  const handleClose = (state: boolean) => {
    setShow(state)
  }

  useEffect(() => {
    if (minDate) {
      setOptions({ ...defaultOptions, minDate: new Date(minDate), maxDate: new Date(minDate) })
    }
  }, [minDate])

  return (
    <Datepicker options={options} onChange={onChange} show={show} setShow={handleClose}>
      <div className='relative'>
        <span className='absolute flex py-[22px] left-3 items-center h-5 w-5 text-[#282929] dark:text-white'>
          <CalendarIcon />
        </span>
        <input
          type='text'
          placeholder='Select Date'
          value={value ? dayjs(value).utc().format('MMMM DD, YYYY') : ''}
          onFocus={() => setShow(!show)}
          className='px-10 py-3 w-full rounded-[42px] bg-white dark:bg-[#1E254E] dark:text-white border-transparent focus:border-gray-500 focus:bg-white focus:ring-0 text-sm'
          readOnly
        />
      </div>
    </Datepicker>
  )
}

export default BasicDatepicker
