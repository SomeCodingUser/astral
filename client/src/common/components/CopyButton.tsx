import { ButtonHTMLAttributes, FC, ReactNode } from 'react'
import { Square2StackIcon } from '@heroicons/react/24/outline'
import toast from 'react-hot-toast'

// common
import { useClipboard } from 'common/hooks/useClipboard'

interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
  value: string
  message?: string
  children?: ReactNode
  buttonClass?: string
  iconClass?: string
}

const CopyButton: FC<Props> = ({
  value,
  children,
  message = 'Copied',
  buttonClass = 'w-full',
  iconClass = 'w-4 h-4',
  ...rest
}) => {
  const { onCopy } = useClipboard(value)

  const handleCopyClick = (): void => {
    onCopy()
    toast.success(message, {
      position: 'bottom-center',
    })
  }

  return (
    <button className={`flex gap-2 ${buttonClass}`} onClick={handleCopyClick} {...rest}>
      {children}
      <Square2StackIcon className={iconClass} stroke='#DE67E4' />
    </button>
  )
}

export default CopyButton
