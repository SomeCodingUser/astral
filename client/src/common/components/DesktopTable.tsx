import { ChevronDownIcon, ChevronUpDownIcon, ChevronUpIcon } from '@heroicons/react/20/solid'
import { Table, flexRender } from '@tanstack/react-table'

interface TableProps<T extends object> {
  table: Table<T>
}

const DesktopTable = <T extends object>({ table }: TableProps<T>) => (
  <table className="min-w-max table-auto font-['Montserrat'] bg-white rounded-[20px] dark:bg-gradient-to-r dark:from-[#4141B3] dark:via-[#6B5ACF] dark:to-[#896BD2] dark:border-none">
    <thead className='text-[#857EC2] text-sm dark:text-white/75 border-b border-gray-200'>
      {table.getHeaderGroups().map((headerGroup) => (
        <tr key={headerGroup.id}>
          {headerGroup.headers.map((header) => (
            <th
              key={header.id}
              {...(header.column.getCanSort()
                ? { onClick: header.column.getToggleSortingHandler() }
                : {})}
              className='px-3 py-4 text-sm font-normal text-start'
            >
              <div className='flex gap-1 justify-items-center align-middle'>
                {header.isPlaceholder
                  ? null
                  : flexRender(header.column.columnDef.header, header.getContext())}
                {header.column.getCanSort() ? (
                  header.column.getIsSorted() === 'asc' ? (
                    <span className='cursor-pointer relative inline-flex items-center rounded-full text-sm font-medium text-[#DE67E4]  focus:z-20 mr-[14px] dark:text-white  dark:border-none'>
                      <span className='sr-only'>Up</span>
                      <ChevronUpIcon className='h-5 w-5' aria-hidden='true' />
                    </span>
                  ) : header.column.getIsSorted() === 'desc' ? (
                    <span className='cursor-pointer relative inline-flex items-center rounded-full text-sm font-medium text-[#DE67E4] focus:z-20 mr-[14px] dark:text-white  dark:border-none'>
                      <span className='sr-only'>Down</span>
                      <ChevronDownIcon className='h-5 w-5' aria-hidden='true' />
                    </span>
                  ) : (
                    <span className='cursor-pointer relative inline-flex items-center rounded-full text-sm font-medium text-[#DE67E4]  focus:z-20 mr-[14px] dark:text-white  dark:border-none'>
                      <span className='sr-only'>UpDown</span>
                      <ChevronUpDownIcon className='h-5 w-5' aria-hidden='true' />
                    </span>
                  )
                ) : (
                  <span></span>
                )}
              </div>
            </th>
          ))}
        </tr>
      ))}
    </thead>
    <tbody className='text-gray-600 text-sm font-light dark:text-white'>
      {table.getRowModel().rows.length ? (
        table.getRowModel().rows.map((row, index) => (
          <tr
            key={row.id}
            className={`hover:bg-gray-100 dark:hover:bg-transparent/10 ${
              index !== table.getRowModel().rows.length - 1 && 'border-y border-gray-200'
            } `}
          >
            {row.getVisibleCells().map((cell) => (
              <td className='whitespace-nowrap px-6 py-4 text-sm font-light' key={cell.id}>
                {flexRender(cell.column.columnDef.cell, cell.getContext())}
              </td>
            ))}
          </tr>
        ))
      ) : (
        <div className='w-full flex p-6 hover:bg-gray-100 dark:hover:bg-transparent/10'>
          <div className='text-[#857EC2] text-sm dark:text-white/75'>No entries to show</div>
        </div>
      )}
    </tbody>
  </table>
)

export default DesktopTable
