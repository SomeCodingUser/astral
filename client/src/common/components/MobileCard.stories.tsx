import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import MobileCard from './MobileCard'

export default {
  title: 'Common/MobileCard',
  component: MobileCard,
} as ComponentMeta<typeof MobileCard>

const Template: ComponentStory<typeof MobileCard> = (args) => <MobileCard {...args} />

export const Primary = Template.bind({})
Primary.args = {
  header: 'Header',
  body: [{ name: 'sample name', value: 'sample value' }],
  id: 'sample id',
}
