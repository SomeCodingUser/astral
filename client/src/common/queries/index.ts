import { gql } from '@apollo/client'

export const QUERY_BLOCK_AND_EXTRINSIC_BY_HASH = gql`
  query BlocksAndExtrinsicByHash($hash: String!) {
    blocks(limit: 10, where: { hash_eq: $hash }) {
      id
      height
    }
    extrinsics(limit: 10, where: { hash_eq: $hash }) {
      id
    }
  }
`
