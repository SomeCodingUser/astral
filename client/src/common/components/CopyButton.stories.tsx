import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import CopyButton from './CopyButton'

export default {
  title: 'Common/CopyButton',
  component: CopyButton,
} as ComponentMeta<typeof CopyButton>

const Template: ComponentStory<typeof CopyButton> = (args) => <CopyButton {...args} />

export const Primary = Template.bind({})
Primary.args = {
  value: 'Value to copy',
}
