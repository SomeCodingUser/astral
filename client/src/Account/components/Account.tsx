import { FC } from 'react'
import { useParams } from 'react-router-dom'
import { useQuery } from '@apollo/client'
import { useErrorHandler } from 'react-error-boundary'

// account
import { QUERY_ACCOUNT_BY_ID, QUERY_LATEST_REWARDS } from 'Account/query'
import {
  AccountDetailsCard,
  AccountExtrinsicList,
  AccountGraphs,
  AccountLatestRewards,
} from 'Account/components'

// layout
import { NotFound } from 'layout/components'

// common
import { Spinner } from 'common/components'
import useMediaQuery from 'common/hooks/useMediaQuery'
import { accountIdToHex, formatAddress } from 'common/helpers/formatAddress'

const Account: FC = () => {
  const { accountId } = useParams<{ accountId?: string }>()

  const convertedAddress = formatAddress(accountId)
  const hexAddress = accountIdToHex(accountId || '')

  const isDesktop = useMediaQuery('(min-width: 1024px)')

  const { data, error, loading } = useQuery(QUERY_ACCOUNT_BY_ID, {
    variables: { accountId: convertedAddress },
  })

  const {
    data: dataRewards,
    error: errorRewards,
    loading: loadingRewards,
  } = useQuery(QUERY_LATEST_REWARDS, {
    variables: { accountId: hexAddress },
  })

  useErrorHandler(error)
  useErrorHandler(errorRewards)

  if (!convertedAddress) {
    return <NotFound />
  }

  if (loading || loadingRewards) {
    return <Spinner />
  }

  if (!data.accountById || !dataRewards.rewardEvents) {
    return <NotFound />
  }

  const account = data.accountById

  return (
    <div className='w-full flex flex-col space-y-4'>
      <AccountDetailsCard
        account={account}
        accountAddress={convertedAddress}
        isDesktop={isDesktop}
      />
      <div className='flex flex-col lg:flex-row lg:justify-between gap-8'>
        <AccountGraphs hexAddress={hexAddress} account={account} isDesktop={isDesktop} />
        <AccountLatestRewards rewards={dataRewards.rewardEvents} isDesktop={isDesktop} />
      </div>
      <AccountExtrinsicList accountId={convertedAddress} />
    </div>
  )
}

export default Account
