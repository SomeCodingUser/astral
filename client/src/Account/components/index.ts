import AccountList from 'Account/components/AccountList'
import Account from 'Account/components/Account'
import AccountDetailsCard from 'Account/components/AccountDetailsCard'
import AccountDetailsTabs from 'Account/components/AccountDetailsTabs'
import AccountBalancePieChart from 'Account/components/AccountBalancePieChart'
import AccountBalanceStats from 'Account/components/AccountBalanceStats'
import AccountExtrinsicList from 'Account/components/AccountExtrinsicList'
import AccountTable from 'Account/components/AccountTable'
import AccountListCard from 'Account/components/AccountListCard'
import AccountLatestRewards from 'Account/components/AccountLatestRewards'
import AccountGraphs from 'Account/components/AccountGraphs'
import AccountGraphTabs from 'Account/components/AccountGraphTabs'
import AccountRewardGraph from 'Account/components/AccountRewardGraph'
import AccountRewardList from 'Account/components/AccountRewardList'
import AccountRewardListCard from 'Account/components/AccountRewardListCard'
import AccountRewardTable from 'Account/components/AccountRewardTable'

export {
  Account,
  AccountList,
  AccountDetailsCard,
  AccountDetailsTabs,
  AccountBalancePieChart,
  AccountBalanceStats,
  AccountExtrinsicList,
  AccountTable,
  AccountListCard,
  AccountLatestRewards,
  AccountGraphs,
  AccountGraphTabs,
  AccountRewardGraph,
  AccountRewardList,
  AccountRewardListCard,
  AccountRewardTable,
}
