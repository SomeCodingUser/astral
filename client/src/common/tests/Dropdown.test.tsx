import { render } from '@testing-library/react'

// common - Dropdown
import { Dropdown } from 'common/components'

describe('Dropdown', () => {
  it('renders properly', async () => {
    const { asFragment } = render(<Dropdown />)

    expect(asFragment()).toMatchSnapshot()
  })
})
