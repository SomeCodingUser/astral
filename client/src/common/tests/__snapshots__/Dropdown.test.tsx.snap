// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders properly 1`] = `
<DocumentFragment>
  <div
    class="relative mt-1 w-36"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="relative w-full cursor-default rounded-lg bg-white py-2 pl-3 pr-10 text-left shadow-md focus:outline-none focus-visible:border-indigo-500 focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75 focus-visible:ring-offset-2 focus-visible:ring-offset-orange-300 sm:text-sm"
      data-headlessui-state=""
      id="headlessui-listbox-button-:r0:"
      type="button"
    >
      <div
        class="flex"
      >
        <svg
          fill="none"
          height="23"
          viewBox="0 0 22 23"
          width="22"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.1425 8.04569H9.57594C9.32825 8.04569 9.12422 8.14841 8.97852 8.33912C8.83281 8.52988 8.75993 8.75001 8.75993 9.01413C8.75993 9.26357 8.83281 9.48369 8.96396 9.65977C9.10967 9.83586 9.29909 9.92389 9.56139 9.92389H12.8401C13.729 9.92389 14.4285 10.2614 14.9385 10.9217C15.4048 11.5233 15.6379 12.257 15.6379 13.1374C15.6379 14.0178 15.4048 14.7515 14.9385 15.3531C14.4285 16.0134 13.729 16.3509 12.8401 16.3509H6.88011V14.1058H12.4613C12.6944 14.1058 12.8838 14.0032 13.0295 13.8124C13.1753 13.6216 13.2335 13.3868 13.2335 13.1227C13.2335 12.8586 13.1607 12.6385 13.0295 12.4624C12.8984 12.2863 12.7089 12.1983 12.4613 12.1983H9.18249C8.2936 12.1983 7.59419 11.8755 7.06959 11.2152C6.60326 10.6282 6.37012 9.89453 6.37012 9.01413C6.37012 8.14841 6.60326 7.42941 7.06959 6.82776C7.59419 6.16749 8.2936 5.82996 9.18249 5.82996H15.1279V8.04569H15.1425Z"
            fill="currentColor"
          />
          <path
            d="M20.5192 5.6826L20.0383 5.93204C19.9655 5.97609 19.9363 6.06413 19.98 6.13749C20.9855 7.9717 21.4372 10.1287 21.1458 12.4177C20.5629 17.0399 16.9345 20.7083 12.3442 21.2953C5.59741 22.1757 -0.100293 16.365 0.89061 9.54173C1.53178 5.11035 4.99993 1.60337 9.38611 0.913719C11.0182 0.664265 12.592 0.79633 14.02 1.25121C14.0055 1.32458 14.0055 1.38327 14.0055 1.45664C14.0055 2.26368 14.7778 2.89465 15.6084 2.6452C15.9873 2.52781 16.2933 2.21966 16.4099 1.83815C16.6576 0.987087 16.031 0.224061 15.2295 0.224061C14.9089 0.224061 14.6175 0.356123 14.3843 0.561553C12.8397 0.0333055 11.1202 -0.142777 9.34239 0.121347C4.57734 0.84035 0.817747 4.64079 0.118288 9.43906C-0.945474 16.8492 5.3205 23.1441 12.6794 22.0436C17.4153 21.3393 21.1749 17.5535 21.8744 12.7846C22.2532 10.2167 21.7578 7.76625 20.6212 5.72665C20.6066 5.6826 20.5629 5.66792 20.5192 5.6826Z"
            fill="currentColor"
          />
        </svg>
        <span
          class="ml-2 block truncate"
        >
          All
        </span>
        <span
          class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-2"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5 text-gray-400 ui-open:rotate-180 ui-open:transform"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M5.23 7.21a.75.75 0 011.06.02L10 11.168l3.71-3.938a.75.75 0 111.08 1.04l-4.25 4.5a.75.75 0 01-1.08 0l-4.25-4.5a.75.75 0 01.02-1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
    </button>
  </div>
</DocumentFragment>
`;
